test/blackbox/run_blackbox_tests/test_cases_key_rotation.c

```diff
@@ -0,0 +1,449 @@
+static void test_case_key_rotation_03(void **state) {
+	execute_test(test_key_rotation_03, state);
+}
+
+/* Test Steps for key rotation Test Case # 3
+
+    Test Steps:
+    1. Open encrypted node instance, rotate it's key with a newkey and close the node.
+    2. Reopen the encrypted node instance with the oldkey
+
+    Expected Result:
+    Opening encrypted node instance should fail when tried to open with oldkey that's
+    been changed to new by key rotate API.
+*/
+static bool test_key_rotation_03(void) {
+	meshlink_destroy("encrypted_conf");
+	meshlink_set_log_cb(NULL, MESHLINK_DEBUG, log_cb);
+
+	// Open a new meshlink instance.
+
+	meshlink_handle_t *mesh = meshlink_open_encrypted("encrypted_conf", "foo", "encrypted", DEV_CLASS_BACKBONE, "oldkey", 6);
+	assert_int_not_equal(mesh, NULL);
+
+	// Key rotate the encrypted_conf storage with new key
+
+	bool keyrotation_status = meshlink_encrypted_key_rotate(mesh, "newkey", 6);
+	assert_int_equal(keyrotation_status, true);
+
+	meshlink_close(mesh);
+
+	// Reopen the meshlink instance with the new key
+
+	mesh = meshlink_open_encrypted("encrypted_conf", "foo", "encrypted", DEV_CLASS_BACKBONE, "oldkey", 6);
+	assert_int_equal(mesh, NULL);
+
+	// Cleanup
+
+	meshlink_destroy("encrypted_conf");
+
+	return true;
+}
+
+/* Execute key rotation Test Case # 4 - Sanity test */
+static void test_case_key_rotation_04(void **state) {
+	execute_test(test_key_rotation_04, state);
+}
+
+/* Test Steps for key rotation Test Case # 4
+    Verify whether key rotation API gracefully handles invitations porting from
+    old key to new key.
+
+    Test Steps:
+    1. Open foo node instance and generate invitations for peer and bar.
+    2. Do key rotation with newkey and verify invitation timestamps post key rotation.
+    3. Change timestamp of peer key to expire and Open instances of foo, bar and peer nodes
+        and try to join bar and peer node.
+
+    Expected Result:
+    Key rotation API should never change the any file status attributes of an invitation file.
+*/
+static bool test_key_rotation_04(void) {
+	meshlink_handle_t *mesh;
+	meshlink_handle_t *mesh1;
+	meshlink_handle_t *mesh2;
+	struct dirent *ent;
+	DIR *d;
+	char invitation_path_buff[500];
+	struct stat invitation_stat;
+	struct stat temp_stat;
+	struct stat peer_stat;
+	struct stat bar_stat;
+	struct utimbuf timebuf;
+	bool join_status;
+
+	meshlink_destroy("encrypted_conf");
+	meshlink_set_log_cb(NULL, MESHLINK_DEBUG, log_cb);
+
+	// Open a new meshlink instance.
+
+	mesh = meshlink_open_encrypted("encrypted_conf", "foo", "encrypted", DEV_CLASS_BACKBONE, "oldkey", 6);
+	assert_int_not_equal(mesh, NULL);
+
+	// Generate invitations
+
+	char *invitation1 = meshlink_invite(mesh, NULL, "peer");
+	assert_int_not_equal(invitation1, NULL);
+
+	// Read the peer invitation file status structure
+
+	strcpy(invitation_path_buff, "encrypted_conf/invitations/");
+	d = opendir(invitation_path_buff);
+	assert(d);
+
+	if(d) {
+		ent = readdir(d);
```

Comment:
There is no guarantee what will be returned by the first call to `readdir()`. It can be `.`, `..` or one of the invitation file names. I checked on my system and different directories result their contents in different orders. So the best thing to do is to convert this into a loop, and do `readdir()`, then `stat()`, then check if `(temp_stat->st_mode & S_IFMT) == S_IFREG`. If that's true, then it's a regular file.